### PR TITLE
feat(share/discovery): Discard method

### LINKED ISSUE #2204: share/discovery: DiscardPeer method
We need a way for the peer manager to report back to discovery if some peer is not useful anymore and to discover someone else instead.

### PR DESCRIPTION
**Pull request by Wondertan**
Fixes #2204 

Regarding the case where we exhausted all the peers on the network and backed them all off. It can happen on the smaller network, and in the worst case, the node will wait 10 minutes(the default backoff time), which is acceptable considering peers from shrexsub #2105

### CODE REVIEW

**Comment by walldiss** [comment]
It seems like with current solution peer could be rediscovered right away. Could we move discarded peer to some set that will prevent it from be in discovery set?

**Comment by walldiss** [comment]
Nvm, answer was in description. I'm a bit sleepy and skipped reading it

**Comment by Wondertan** [comment]
...
```
=== RUN   TestIntegration/get_peer_from_discovery
    manager_test.go:446: 
        	Error Trace:	/home/runner/work/celestia-node/celestia-node/share/p2p/peers/manager_test.go:446
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	Test:       	TestIntegration/get_peer_from_discovery
```

**Review by walldiss — APPROVED** [review]

**Comment by distractedm1nd** [comment]
For the integration it may be easier if the Discard API returns a boolean, if it was in discovery or not. This is because:

For removing a peer from the pool, we rely on discovery to call our callback, which calls pool.remove
If we call disc.Discard(peer), it will call pool.remove through the callback
But not all peers are from discovery, so if Discard didnt get rid of the peer we need to call pool.remove ourselves

Not ideal but its a solution

File: share/availability/discovery/discovery.go [hunk 1]
```diff
@@ -122,6 +122,30 @@ func (d *Discovery) Peers(ctx context.Context) ([]peer.ID, error) {
...
+	d.onUpdatedPeers(id, false)
+	log.Debugw("removed peer from the peer set", "peer", id)
+
+	if d.set.Size() < d.set.Limit() {
+		// trigger discovery
+		select {
+		case d.triggerDisc <- struct{}{}:
+		default:
+		}
+	}
```
Comment (renaynay) [inline share/availability/discovery/discovery.go:144]: could make a method to DRY this but nit

**Review by renaynay — APPROVED** [review]

File: share/availability/discovery/discovery.go (hunk 1 above)
Comment (Wondertan) [reply share/availability/discovery/discovery.go:144]: I would, but it is only used in one place.

**Review by distractedm1nd — APPROVED** [review]

**Pull request merged**